tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+{
+	struct channel *chan = CONTAINER_OF(l2cap_chan, struct channel, le);
+
+	LOG_DBG("chan_id: %d", chan->chan_id);
+}
+
+static void chan_reconfigured_cb(struct bt_l2cap_chan *l2cap_chan)
+{
+	struct channel *chan = CONTAINER_OF(l2cap_chan, struct channel, le);
+
+	LOG_DBG("chan_id: %d", chan->chan_id);
+}
+
+static const struct bt_l2cap_chan_ops l2cap_ops = {
+	.alloc_buf = chan_alloc_buf_cb,
+	.recv = chan_recv_cb,
+	.sent = chan_sent_cb,
+	.connected = chan_connected_cb,
+	.disconnected = chan_disconnected_cb,
+	.status = chan_status_cb,
+	.released = chan_released_cb,
+	.reconfigured = chan_reconfigured_cb,
+};
+
+static void connect_num_channels(uint8_t num_l2cap_channels)
+{
+	struct channel *chan = NULL;
+
+	#define L2CAP_ECRED_CHAN_MAX 5
+	struct bt_l2cap_chan *allocated_channels[L2CAP_ECRED_CHAN_MAX] = { NULL };
+	#undef L2CAP_ECRED_CHAN_MAX
+	int err = 0;
+
+	for (int i = 0; i < num_l2cap_channels; i++) {
+		chan = get_free_channel();
+		if (!chan) {
+			FAIL("failed, chan not free");
+			return;
+		}
+		chan->le.chan.ops = &l2cap_ops;
+		chan->le.rx.mtu = DATA_MTU;
+		chan->le.rx.mps = DATA_MPS;
+		allocated_channels[i] = &chan->le.chan;
+	}
+
+	err = bt_l2cap_ecred_chan_connect(default_conn, allocated_channels, servers[0].psm);
+	if (err) {
+		FAIL("can't connect ecred %d ", err);
+	}
+}
+
+static void disconnect_all_channels(void)
+{
+	int err = 0;
+
+	for (int i = 0; i < ARRAY_SIZE(channels); i++) {
+		if (channels[i].in_use) {
+			LOG_DBG("Disconnecting channel: %d)", channels[i].chan_id);
+			err = bt_l2cap_chan_disconnect(&channels[i].le.chan);
+			if (err) {
+				LOG_DBG("can't disconnnect channel (err: %d)", err);
+			}
+			channels[i].in_use = false;
+		}
+	}
+}
+
+static int accept(struct bt_conn *conn, struct bt_l2cap_chan **l2cap_chan)
+{
+	struct channel *chan;
+
+	chan = get_free_channel();
+	if (!chan) {
+		return -ENOMEM;
+	}
+
+	chan->le.chan.ops = &l2cap_ops;
+	chan->le.tx.mtu = DATA_MTU;
+	chan->le.rx.mtu = DATA_MTU;
+
+	*l2cap_chan = &chan->le.chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server *get_free_server(void)
+{
+	for (int i = 0U; i < SERVERS; i++) {
+		if (servers[i].psm) {
+			continue;
+		}
+
+		return &servers[i];
+	}
+
+	return NULL;
+}
+
+static void register_l2cap_server(void)
+{
+	struct bt_l2cap_server *server;
+
+	server = get_free_server();
+	if (!server) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	server->accept = accept;
+	server->psm = 0;
+
+	if (bt_l2cap_server_register(server) < 0) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	LOG_DBG("L2CAP server registered, PSM:0x%X", server->psm);
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
```

Comment:
Why are we taking a reference here?

Reply:
Looks correct to me. Only way to avoid the `ref()` is to remove the `default_conn`-based connection tracking completely.

Reply:
The only thing that might need fixing is checking if `default_conn` is already set, but in the case that this code supports multiple connections then the need for a `default_conn` variable seems questionable.

Reply:
for now it will not support multiple connections... can be postponed for the future integration

Reply:
> Looks correct to me. Only way to avoid the `ref()` is to remove the `default_conn`-based connection tracking completely.

But is it necessary? We don't do this for the majority of samples and other tests.

Reply:
What exactly are you referring to? The existence of the `default_conn` variable, or the `ref()` in the connected callback. The application doesn't own a reference for the connection given as a parameter to the connected callback, so it's a bug if it stores a pointer to it without incrementing the reference count.

Reply:
> What exactly are you referring to? The existence of the `default_conn` variable, or the `ref()` in the connected callback? The application doesn't own a reference for the connection given as a parameter to the connected callback, so it's a bug if it stores a pointer to it without incrementing the reference count.

Oh? Then we indeed have that bug multiple places in the code base (mostly code introduced by me I think) - I'll fix that :)